src/Utilities/GridFileReader.f90

```diff
@@ -0,0 +1,521 @@
+      call store_error_unit(this%inunit)
+    end if
+    call urword(line, lloc, istart, istop, 1, ival, rval, 0, 0)
+    this%grid_type = line(istart:istop)
+    call upcase(this%grid_type)
+
+    ! version
+    read (this%inunit) line
+    lloc = 1
+    call urword(line, lloc, istart, istop, 0, ival, rval, 0, 0)
+    call urword(line, lloc, istart, istop, 2, ival, rval, 0, 0)
+    this%version = ival
+
+    ! ntxt
+    read (this%inunit) line
+    lloc = 1
+    call urword(line, lloc, istart, istop, 0, ival, rval, 0, 0)
+    call urword(line, lloc, istart, istop, 2, ival, rval, 0, 0)
+    this%ntxt = ival
+
+    ! lentxt
+    read (this%inunit) line
+    lloc = 1
+    call urword(line, lloc, istart, istop, 0, ival, rval, 0, 0)
+    call urword(line, lloc, istart, istop, 2, ival, rval, 0, 0)
+    this%lentxt = ival
+
+  end subroutine read_header_meta
+
+  !> @brief Read the header body section (text following the first
+  !< four "meta" lines). Internal use only.
+  subroutine read_header_body(this)
+    ! dummy
+    class(GridFileReaderType) :: this
+    ! local
+    character(len=:), allocatable :: body
+    character(len=:), allocatable :: line
+    character(len=10) :: key, dtype
+    real(DP) :: rval
+    integer(I4B) :: i, lloc, istart, istop, ival
+    integer(I4B) :: nvars, ndim, dim, ishp
+    integer(I4B), allocatable :: shp(:)
+
+    allocate (this%keys(this%ntxt))
+    allocate (character(len=this%lentxt*this%ntxt) :: body)
+    allocate (character(len=this%lentxt) :: line)
+
+    nvars = 0
+    read (this%inunit) body
+    do i = 1, this%lentxt * this%ntxt, this%lentxt
+      line = body(i:i + this%lentxt - 1)
+      lloc = 1
+
+      ! key
+      lloc = 1
+      call urword(line, lloc, istart, istop, 1, ival, rval, 0, 0)
+      key = line(istart:istop)
+      call upcase(key)
+      nvars = nvars + 1
+      this%keys(nvars) = key
+
+      ! type
+      call urword(line, lloc, istart, istop, 1, ival, rval, 0, 0)
+      dtype = line(istart:istop)
+      call upcase(dtype)
+      if (dtype == "INTEGER") then
+        call this%typ%add(key, 1)
+      else if (dtype == "DOUBLE") then
+        call this%typ%add(key, 2)
+      end if
+
+      ! dims
+      call urword(line, lloc, istart, istop, 0, ival, rval, 0, 0)
+      call urword(line, lloc, istart, istop, 2, ival, rval, 0, 0)
+      ndim = ival
+      call this%dim%add(key, ndim)
+
+      ! shape
+      if (allocated(shp)) deallocate (shp)
+      allocate (shp(ndim))
+      if (ndim > 0) then
+        do dim = 1, ndim
+          call urword(line, lloc, istart, istop, 2, ival, rval, 0, 0)
+          shp(dim) = ival
+        end do
+        ishp = size(this%shp)
+        call ExpandArray(this%shp, increment=ndim)
+        this%shp(ishp + 1:ishp + ndim) = shp
+        call this%shp_idx%add(key, ishp + 1)
+      end if
+    end do
+  end subroutine read_header_body
+
+  !> @brief Skim the file and build an index. Internal use only.
```

Comment:
I guess to get the offsets, this could just multiply shape by byte count instead of reading the file.